Implement Acknowledge Feature in Modules/Forms #25

MasterHans opened the pull request:
Added 'acknowledge' => true feature to sveral forms:
Forms for acknowledge :

1) http://humhub.localhost/index.php?r=admin%2Fauthentication
2) http://humhub.localhost/index.php?r=admin%2Fgroup%2Fedit&id=1
3) http://humhub.localhost/index.php?r=admin%2Fsetting%2Fbasic
4) http://humhub.localhost/index.php?r=admin%2Fsetting%2Fdesign
5) http://humhub.localhost/index.php?r=activity%2Fadmin%2Fdefaults
6) http://humhub.localhost/index.php?r=notification%2Fadmin%2Fdefaults
7) http://humhub.localhost/index.php?r=admin%2Fsetting%2Fcaching
8) http://humhub.localhost/index.php?r=admin%2Fsetting%2Ffile
9) http://humhub.localhost/index.php?r=admin%2Fsetting%2Fmailing-server
10) http://humhub.localhost/index.php?r=admin%2Fsetting%2Fproxy
11) http://humhub.localhost/index.php?r=admin%2Fsetting%2Fstatistic
12) http://humhub.localhost/index.php?r=admin%2Fsetting%2Foembed-edit

luke- commented:
@MasterHans  Can you please check the PR? It should go against the Develop Branch. There are also some irrelevant commits here.

MasterHans commented:
Hi, guys!

I have added **'acknowledge' => true** feature but haven't finished yet for all core forms. I will continue, but I have a question.
For forms like this:
**http://humhub.localhost/index.php?r=admin%2Fsetting%2Fproxy**
 I have created **humhub\compat\CAcknowledgeActiveForm** class
It is a copy of  **humhub\compat\CActiveForm**, witch extened **humhub\modules\ui\form\widgets\ActiveForm**;

And then I added this code to the form which using  **humhub\compat\CActiveForm**.

There are several other forms where I plan to make the same implementation.

Am I right?

MasterHans commented:
> @MasterHans Can you please check the PR? It should go against the Develop Branch. There are also some irrelevant commits here.

@luke-  I have changed the base from master to develop. Is that enough or it has some other issues?

luke- commented:
@MasterHans ok, looks better now!

We actually want to get rid of this CActiveForm usage completely if possible (as it is deprecated).
Can you please migrate the CActiveForms in the Core to the normal new ui/ActiveForm?

MasterHans commented:
> @MasterHans ok, looks better now!
> 
> We actually want to get rid of this CActiveForm usage completely if possible (as it is deprecated).
> Can you please migrate the CActiveForms in the Core to the normal new ui/ActiveForm?

Ok, should I raise another issue?  Or make it in this PR?

ArchBlood commented:
Also noticed that `CHtml` is being used should this also be migrated?

luke- commented:
@ArchBlood Good point!

Yes, we should also migrate all CHTML calls directly if possible.

luke- commented:
@MasterHans You can do this in this PR.

luke- reviewed (changes requested):
Thanks, please see my review comments.

If you're reworking the views, can you please also add available variables to the view header:
Example: https://github.com/humhub/humhub/blob/master/protected/humhub/modules/admin/views/module/list.php#L5-L9

If you see any other bad code style e.g. https://github.com/humhub/humhub/pull/4776/files#diff-ca41b3456a6dfa34bc1fa9e02ec2f3452ae06fb5fd524202b2ca731174133063R23 No `Html` Helper used, feel free to fix this also.

MasterHans commented:
> Thanks, please see my review comments.
> 
> If you're reworking the views, can you please also add available variables to the view header:
> Example: https://github.com/humhub/humhub/blob/master/protected/humhub/modules/admin/views/module/list.php#L5-L9
> 
> If you see any other bad code style e.g. https://github.com/humhub/humhub/pull/4776/files#diff-ca41b3456a6dfa34bc1fa9e02ec2f3452ae06fb5fd524202b2ca731174133063R23 No `Html` Helper used, feel free to fix this also.

Done!

MasterHans commented on protected/humhub/modules/admin/views/group/edit.php:
```diff
@@ -1,17 +1,17 @@
 ?>
 
-<?php $this->beginContent('@admin/views/group/_manageLayout.php', ['group' => $group]) ?>
+<?php $this->beginContent('@admin/views/group/_manageLayout.php', ['group' => $group, 'acknowledge' => true]) ?>
```
Why not? I think that user can leave unsaved changes on the page and move out of the page

yurabakhtin replied:
I agree with @MasterHans

luke- replied:
@yurabakhtin Are you sure? This line introduces the variable `acknowledge` in this view 
https://github.com/humhub/humhub/blob/29df6ab189f83b7c554a3b7fa3308085e3215bdd/protected/humhub/modules/admin/views/group/_manageLayout.php#L8 without using it.

yurabakhtin replied:
I was wrong, I have reviewed and got we should pass the param into the `ActiveForm` instead of `beginContent` as it is already implemented in the latest version of this PR, so this conversation can be resolved.

MasterHans commented on protected/humhub/modules/admin/views/group/members.php:
```diff
@@ -24,7 +24,7 @@
      <div class="row">
         <div class="form-group col-md-6">
-            <?php $form = ActiveForm::begin(['action' => ['/admin/group/add-members']]); ?>
+            <?php $form = ActiveForm::begin(['action' => ['/admin/group/add-members'], 'acknowledge' => true]); ?>
```
done!

MasterHans commented on protected/humhub/modules/admin/views/setting/file.php:
```diff
@@ -7,57 +7,41 @@
 
 <?= $form->errorSummary($model); ?>
 
 <div class="form-group">
```
Done!

luke- replied:
@MasterHans The `<div class="form-group">` line 23, is still there. These divs are automatically generated by the Bootstrap ActiveForm. Please remove this. Can you please also move the `hint` from the view in the `attributeHints` section of the form model.

MasterHans commented on protected/humhub/modules/admin/views/setting/logs.php:
```diff
@@ -1,32 +1,32 @@
 
 <?= $form->errorSummary($model); ?>
 
 <div class="form-group">
```
Done!

MasterHans commented on protected/humhub/modules/admin/views/setting/oembed_edit.php:
```diff
@@ -24,26 +23,24 @@
     <p class="help-block"><?= Yii::t('AdminModule.settings', 'Url Prefix without http:// or https:// (e.g. youtube.com)'); ?></p>
 </div>
 
 <div class="form-group">
```
Done!

luke- replied:
@MasterHans The `<div class="form-group">` line 31, these divs are automatically generated by the Bootstrap ActiveForm. Please remove this. Can you please also move the `hint` from the view in the `attributeHints` section of the form model.

MasterHans commented on protected/humhub/modules/admin/views/setting/proxy.php:
```diff
@@ -1,60 +1,53 @@
 
 <?= $form->errorSummary($model); ?>
 
 <div class="form-group">
```
Done!

yurabakhtin replied:
@luke- Why is it not required? I think it is strange why all similar forms under tab "Administration" -> "Settings" -> "Advanced" require the confirmation to leave a page with changes but the page "Proxy" doesn't require.

Also the form "Logs" doesn't have the param `['acknowledge' => true]`.

luke- replied:
@yurabakhtin Ok that's right, my main concern was the logs in the information section. There's nothing wrong with leaving it in here to stay consistent.

MasterHans commented:
All Done with these forms:
1) http://humhub.localhost/index.php?r=admin%2Fauthentication
2) http://humhub.localhost/index.php?r=admin%2Fgroup%2Fedit&id=1
3) http://humhub.localhost/index.php?r=admin%2Fsetting%2Fbasic
4) http://humhub.localhost/index.php?r=admin%2Fsetting%2Fdesign
5) http://humhub.localhost/index.php?r=activity%2Fadmin%2Fdefaults
6) http://humhub.localhost/index.php?r=notification%2Fadmin%2Fdefaults
7) http://humhub.localhost/index.php?r=admin%2Fsetting%2Fcaching
8) http://humhub.localhost/index.php?r=admin%2Fsetting%2Ffile
9) http://humhub.localhost/index.php?r=admin%2Fsetting%2Fmailing-server
10) http://humhub.localhost/index.php?r=admin%2Fsetting%2Fproxy
11) http://humhub.localhost/index.php?r=admin%2Fsetting%2Fstatistic
12) http://humhub.localhost/index.php?r=admin%2Fsetting%2Foembed-edit
13) http://humhub.localhost/index.php?r=admin%2Fuser%2Fadd
14) http://humhub.localhost/index.php?r=admin%2Fuser-profile%2Fedit-category&id=1

15) http://humhub.localhost/index.php?r=space%2Fmanage&cguid=534be4e8-e0ef-43d1-a794-164352751ce5
16) http://humhub.localhost/index.php?r=space%2Fmanage%2Fdefault%2Fadvanced&cguid=534be4e8-e0ef-43d1-a794-164352751ce5
17) http://humhub.localhost/index.php?r=admin%2Fspace%2Fsettings

20) http://humhub.localhost/index.php?r=user%2Faccount%2Fedit
21) http://humhub.localhost/index.php?r=activity%2Fuser
22) http://humhub.localhost/index.php?r=notification%2Fuser
23) http://humhub.localhost/index.php?r=user%2Faccount%2Fedit-settings
24) http://humhub.localhost/index.php?r=user%2Faccount%2Fchange-username
25) http://humhub.localhost/index.php?r=user%2Faccount%2Fchange-email
26) http://humhub.localhost/index.php?r=user%2Faccount%2Fchange-password

MasterHans commented:
I have fixed the issue with http://humhub.localhost/user/account/edit
I remind you what it was:
It triggers the acknowledge dialog without changing.

The serializing of the form cause the issue.
This string causes the issue:

```
var serializeFormState = function($form) {	    var serializeFormState = function ($form) {
        return $form.find(':not([data-prevent-statechange])').serialize();	       
    };
```

It doubles the input elements like this:
_csrf=-01WlNsGyHVxoaxw3A-k0NFX8ZjMrk8ij9G1yJ9EVJSjLz7CilL5DxbH4wXvX-iClxyktficN3PJlOeapzAM3Q%3D%3D&
Profile%5Bfirstname%5D=Aleksandr&Profile%5Blastname%5D=Suvorov&Profile%5Btitle%5D=System%20Administration&Profile%5Bgender%5D=&Profile%5Bstreet%5D=&Profile%5Bzip%5D=&Profile%5Bcity%5D=&Profile%5Bcountry%5D=&Profile%5Bstate%5D=&Profile%5Bbirthday%5D=&Profile%5Babout%5D=&
Profile%5Bfirstname%5D=Aleksandr&Profile%5Blastname%5D=Suvorov&Profile%5Btitle%5D=System%20Administration&Profile%5Bgender%5D=&Profile%5Bstreet%5D=&Profile%5Bzip%5D=&Profile%5Bcity%5D=&Profile%5Bcountry%5D=&Profile%5Bstate%5D=&Profile%5Bbirthday%5D=&Profile%5Babout%5D=&

Profile%5Bphone_private%5D=&Profile%5Bphone_work%5D=&Profile%5Bmobile%5D=&Profile%5Bfax%5D=&Profile%5Bim_skype%5D=&Profile%5Bim_xmpp%5D=&
Profile%5Bphone_private%5D=&Profile%5Bphone_work%5D=&Profile%5Bmobile%5D=&Profile%5Bfax%5D=&Profile%5Bim_skype%5D=&Profile%5Bim_xmpp%5D=&

Profile%5Burl%5D=&Profile%5Burl_facebook%5D=&Profile%5Burl_linkedin%5D=&Profile%5Burl_xing%5D=&Profile%5Burl_youtube%5D=&Profile%5Burl_vimeo%5D=&Profile%5Burl_flickr%5D=&Profile%5Burl_myspace%5D=&Profile%5Burl_twitter%5D=&
Profile%5Burl%5D=&Profile%5Burl_facebook%5D=&Profile%5Burl_linkedin%5D=&Profile%5Burl_xing%5D=&Profile%5Burl_youtube%5D=&Profile%5Burl_vimeo%5D=&Profile%5Burl_flickr%5D=&Profile%5Burl_myspace%5D=&Profile%5Burl_twitter%5D=


Looks like you can not use :not without any selector.

I have changed it into 

`$form.find('input, select, textarea').not('[data-prevent-statechange]').serialize();`

MasterHans commented:
@buddh4 looks like richtext teaxarea working now
http://humhub.localhost/index.php?r=space%2Fmanage&cguid=534be4e8-e0ef-43d1-a794-164352751ce5

MasterHans commented:
@luke- 
http://humhub.localhost/index.php?r=admin%2Fsetting%2Fdesign
The image uploading did not trigger the warning message. Should I fix it? Or maybe this is not necessary?

MasterHans commented:
@luke- I have finished with testing. 
Ready to proceed to modules.

I mentioned an acceptance test 
`The option is unfortunately not documented, can you please add PHPDocs here and an Acceptance Test:`

in connection with ActiveForm option.
Please describe what this test should do?
Enter some form make changes and switch to another tab. And then look for the warning message?

buddh4 commented:
I think we should do acceptance tests for such components in the devtools module in the future.

MasterHans commented:
@luke-  I have finished with this PR. I think we can merge it.

Next step:
1) I will create PRs for each modules... I will Migrate CAtiveForm to ActiveForm and CHTML as well.
2) Add acceptance test in DepTool module.

luke- commented:
@MasterHans Ok thanks! I'll test it in the next days. We will merge this into the version 1.9 - as we are already doing a feature freeze for version 1.8.

ArchBlood commented on protected/config/common.php:
```diff
@@ -6,4 +6,10 @@
+    'components' => [
+        'urlManager' => [
+            'showScriptName' => false,
+            'enablePrettyUrl' => true,
```
Just to add a note here, if this is set to true then you'll also have to edit the `.htaccess.dist` file and change the name as well to reflect it else you'll end up with a frontend server error

https://github.com/humhub/humhub/blob/master/.htaccess.dist

Personally I think adding this here is a good thing, only thing I personally think is that it should be commented out or at least have `enablePrettyUrl` set to `false` if the above isn't changed

luke- replied:
@ArchBlood Thanks for the hin. The common.php  change should not longer part of this PR.

yurabakhtin commented:
@luke- I reviewed this PR as you requested, I think it works well, but for me it is not clear why some forms should not have the "acknowledge" feature, e.g. I wrote above about pages "Proxy" and "Logs".

Also why the same confirmation is not used on page refresh or on close the page/tab/window?
It seems standard browser conformation dialog is used in such cases instead of custom humhub with text "There are unsaved changes, do you really want to leave this page?":

<img width="1148" alt="different_confirmation" src="https://user-images.githubusercontent.com/6995524/107476590-35a9a780-6b87-11eb-914d-51abb37e21b8.png">

luke- commented:
@yurabakhtin Thanks for the review. If the browsers already provide a confirmation dialog for the reload case, then the additional HumHub Confirm is not needed here? Or do you see advantages if we replace the browser dialog with the HumHub dialog here?

yurabakhtin commented:
@luke- I think it would looks better if in all cases we have the same confirmation dialog.

luke- commented:
@yurabakhtin Ok, can you please create a new issue for this. e.g. something like "ActiveForm Acknloge should so handle Reload Cases".

yurabakhtin commented:
@luke- I have created that here https://github.com/humhub/humhub/issues/4866

MasterHans commented:
@luke-  Should I do something else here?

luke- commented:
@MasterHans I think it looks ok now. We'll postpone it into v1.9
